service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,164 @@
+package com.epam.esm.service.impl;
+
+import com.epam.esm.converter.GiftCertificateDTOToEntityConverter;
+import com.epam.esm.converter.GiftCertificateEntityToDTOConverter;
+import com.epam.esm.dto.GiftCertificateDTO;
+import com.epam.esm.exception.ServiceException;
+import com.epam.esm.hibernate.CertificateRepository;
+import com.epam.esm.persistence.GiftCertificateEntity;
+import com.epam.esm.persistence.TagEntity;
+import com.epam.esm.service.GiftCertificateService;
+import com.epam.esm.validator.PreparedValidatorChain;
+import com.epam.esm.validator.realisation.IntermediateCertificateLink;
+import com.epam.esm.validator.realisation.IntermediateSortLink;
+import com.epam.esm.validator.realisation.certificate.*;
+import com.epam.esm.validator.realisation.sort.FieldValidatorLink;
+import com.epam.esm.validator.realisation.sort.MethodValidatorLink;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Service
+public class GiftCertificateServiceImpl implements GiftCertificateService {
+
+    private final GiftCertificateEntityToDTOConverter converterToDto;
+    private final GiftCertificateDTOToEntityConverter converterToEntity;
+    private final CertificateRepository repository;
+
+    @Autowired
+    public GiftCertificateServiceImpl(GiftCertificateEntityToDTOConverter converterToDto,
+                                      GiftCertificateDTOToEntityConverter converterToEntity,
+                                      CertificateRepository repository) {
+        this.converterToDto = converterToDto;
+        this.converterToEntity = converterToEntity;
+        this.repository = repository;
+    }
+
+    @Override
+    public List<GiftCertificateDTO> findAll(int limit, int page) {
+        return createResultList(repository.findAll(limit, page));
+    }
+
+    @Override
+    public Optional<GiftCertificateDTO> find(long id) {
+        Optional<GiftCertificateEntity> certificateFromDao = repository.find(id);
```

Comment:
I would rather prefer to immediately throw an Exception (some sort of NotFoundException) on service layer when the entity is not found. And the use @ExceptionHandler method to catch and then compose and build error response.
I don't like this chain of Optionals in returning values.